rfcs/text/0000-multiple-users.md

```diff
@@ -0,0 +1,419 @@
+root of the event. Typically a privilege change event will already have happened,
+for example: bob "su" as root; and subsequent events will show the root user
+performing the actions.
+
+### Identity and Access Management
+
+Whenever a user is performing an action that affects another user -- typically
+in IAM scenarios -- the user affected by the action is captured at
+`user.target`. The user performing the IAM activity is captured at the root
+of the event.
+
+Examples of IAM activity include:
+
+* user-a creates or deletes user-b
+* user-a modifies user-b
+
+In the create/delete scenarios, there's either no prior state (user creation)
+or no post state (user deletion). In these cases, only `user` at the root and
+`user.target` must be populated.
+
+Example where "root" creates user "bob":
+
+```JSON
+{
+  "user": {
+    "name": "root",
+    "id": "1",
+    "target": {
+      "name": "bob",
+      "id": "1002",
+      ...
+    }
+  }
+}
+```
+
+When there's a change of state to an existing user, `user.target` must be used
+to capture the prior state of the user, and `user.changes` should list only
+the changes that were performed.
+
+Example where "root" renames user "bob" to "bob.barker":
+
+```JSON
+{
+  "user": {
+    "name": "root",
+    "id": "1",
+    "target": {
+      "name": "bob",
+      "id": "1002"
+    },
+    "changes": {
+      "name": "bob.barker"
+    }
+  }
+}
+```
+
+You'll note in the example above that the user ID is not repeated under `user.changes.*`,
+since the ID didn't change.
+
+### Combining IAM and Privilege Escalation
+
+We've covered above how `user.target` and `user.changes` can be used at the same time.
+If privilege escalation is captured in the same IAM event, `user.effective`
+should of course be used as well.
+
+Here's the "rename" example from the IAM section above. In the following example,
+we know "alice" is escalating privileges as "root", in order to modify user "bob":
+
+```JSON
+{
+  "user": {
+    "name": "alice",
+    "id": "1001",
+    "effective": {
+      "name": "root",
+      "id": "1"
+    },
+    "target": {
+      "name": "bob",
+      "id": "1002"
+    },
+    "changes": {
+      "name": "bob.barker"
+    }
+  }
+}
+```
+
+### Pivoting via related.user
+
+Any event that has user(s) in it should always populate the array field `related.user`
+with all usernames seen on the event. Note that this field is not a nesting of
+all user fields, it's a flat array meant to contain user identifiers.
+
+Taking the example from `user.changes` again, and populating `related.user` as well,
+the event now looks like:
+
+```JSON
+{
+  "user": {
+    "name": "alice",
+    "id": "1001",
+    "effective": {
+      "name": "root",
+      "id": "1"
+    },
+    "target": {
+      "name": "bob",
+      "id": "1002"
+    },
+    "changes": {
+      "name": "bob.barker"
+    }
+  },
+  "related": { "user": ["alice", "root", "bob", "bob.barker"] }
```

Comment:
Apologies if this has already been addressed, but one concern that was raised by the detections team was the uniqueness of these names. If `related.user` is intended to be used programmatically, my (naive) sense is that `id` would be a better candidate here.

Reply:
Yes, this has been raised in the past. We haven't addressed it directly anywhere yet, however. One potential solution that has been floated would be that `related.user` could perhaps contain all seen usernames + user IDs.

The field is only meant to pivot and find a user, no matter where in the event they appeared. It's not meant to have semantic meaning.

This RFC is not meant to address this issue, however. We will tackle that issue separately.